### PR TITLE
Don't NFW for expected exceptions that IL Spy throws

### PR DESCRIPTION
**Pull request by davidwengier**
Fixes [AB#1544391](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1544391)

The thrown exceptions in question:
https://github.com/icsharpcode/ILSpy/blob/master/ICSharpCode.Decompiler/CSharp/CSharpDecompiler.cs#L953-L956

### CODE REVIEW

File: src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs [hunk 1]
```diff
@@ -103,14 +103,26 @@ public static async Task<Document> FormatDocumentAsync(Document document, Syntax
...
-            text += "#endif" + Environment.NewLine;
+                return document.WithText(SourceText.From(text));
+            }
+            catch (NotSupportedException)
+            {
+                // IL Spy throws for "Decompiling types that are not part of the main module is not supported."
+            }
+            catch (InvalidOperationException)
+            {
+                // IL Spy throws if it can't find the type name in its type system.
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs:122]: Why would this happen though? Wouldn't this indicate a bug on our side?

File: src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs [hunk 2]
```diff
@@ -103,14 +103,26 @@ public static async Task<Document> FormatDocumentAsync(Document document, Syntax
...
-            text += "#if false // " + CSharpEditorResources.Decompilation_log + Environment.NewLine;
-            text += logger.ToString();
-            text += "#endif" + Environment.NewLine;
+                return document.WithText(SourceText.From(text));
+            }
+            catch (NotSupportedException)
+            {
+                // IL Spy throws for "Decompiling types that are not part of the main module is not supported."
+            }
+            catch (InvalidOperationException)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs:120]: Should this be filtering on the exception message? I recognize that's potentially sketchy but I worry that catching such broad exceptions is also problematic.

File: src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs [hunk 3]
```diff
@@ -103,14 +103,26 @@ public static async Task<Document> FormatDocumentAsync(Document document, Syntax
...
+                text += "#endif" + Environment.NewLine;
 
-            text += "#if false // " + CSharpEditorResources.Decompilation_log + Environment.NewLine;
-            text += logger.ToString();
-            text += "#endif" + Environment.NewLine;
+                return document.WithText(SourceText.From(text));
+            }
+            catch (NotSupportedException)
+            {
+                // IL Spy throws for "Decompiling types that are not part of the main module is not supported."
```
Comment (jasonmalinowski) [inline src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs:118]: Should we just not call them for these in the first place?

File: src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs (hunk 1 above)
Comment (davidwengier) [reply src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs:122]: No idea, but given we have a fallback for creating source from metadata, I don't think the NFW is helping us in any way regardless.

File: src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs (hunk 3 above)
Comment (davidwengier) [reply src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs:118]: Yeah, I thought about that, but that just means copying some logic from ILSpy into here, which seems fragile. As per the above, I don't think an NFW when ILSpy can't do something is helping us in any way

File: src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs (hunk 2 above)
Comment (davidwengier) [reply src/EditorFeatures/CSharp/DecompiledSource/CSharpDecompiledSourceService.cs:120]: The rest of the code here is string concatenation and ToStringing a string builder. I can isolate the catch to just the ILSpy call if that makes you feel better?

**Comment by CyrusNajmabadi** [comment]
What's your current thinking on the best way forward here?  Can we get a custom exception on their side for 'expected' failures?  InvalidOp def feels more like they're crashing, and i'd rather redirect information tehir way so they can fix that :)

**Comment by davidwengier** [comment]
Jason and I chatted about this, and I opened a discussion on their repo for this. The Prism bug was just for the NotSupportedException, the Invalid Op was just me being thorough, but Jason convinced me it was wrong

**Comment by jasonmalinowski** [comment]
The InvalidOperationException we think might have actually been our fault handing in the wrong data so we'll try to fix that bug. Generally, even the NotSupportedException is a customer-impacting issue, we think it's just a low-priority issue since it only impacts very few customers. But removing the telemetry means if it suddenly impacts a _lot_ of customers, we won't know anymore.

**Pull request closed without merging**